ensime_shared/ensime.py

```diff
@@ -136,16 +136,23 @@ def queue_poll(self, sleep_t=0.5):
 
         Value of sleep is low to improve responsiveness.
         """
+        connection_alive = True
         while self.running:
             if self.ws:
                 def logger_and_close(m):
                     self.log("Websocket exception: {}".format(m))
-                    self.teardown()
+                    if not self.running:
```

Comment:
Yup, that makes sense to me. 

The code doesn't reflect that approach though, because we still fire `logger_and_close()` on any exception, which in turn is guaranteed to call `teardown()` the first time (because we're still `running`), which is guaranteed to set `self.running = False`, which in turn ends the loop and complete's the thread. 

For the approach you're talking about- which I'm all for - I think we need to be more selective about which exceptions we catch with `teardown()`.